substrate/frame/nomination-pools/src/migration.rs

```diff
@@ -107,6 +107,109 @@ pub mod unversioned {
 			Ok(())
 		}
 	}
+
+	/// Migrate existing pools from [`adapter::StakeStrategyType::Transfer`] to
+	/// [`adapter::StakeStrategyType::Delegate`].
+	///
+	/// Note: This only migrates the pools, the members are not migrated. They can use the
+	/// permission-less [`Pallet::claim_delegation()`] to migrate their funds.
+	///
+	/// This migration does not break any existing pool storage item, does not need to happen in any
+	/// sequence and hence can be applied unversioned on a production runtime.
+	pub struct DelegationStakeMigration<T>(sp_std::marker::PhantomData<T>);
+
+	impl<T: Config> OnRuntimeUpgrade for DelegationStakeMigration<T> {
+		fn on_runtime_upgrade() -> Weight {
+			if StakeStrategyType::<T>::get() == adapter::StakeStrategyType::Transfer {
+				return T::DbWeight::get().reads_writes(1, 0)
+			}
+
+			let mut migrate_count: u32 = 0;
+			let mut fail_count: u32 = 0;
+			BondedPools::<T>::iter().for_each(|(id, _inner)| {
+				if Pallet::<T>::migrate_to_delegate_stake(id).is_ok() {
+					migrate_count.saturating_inc();
+				} else {
+					fail_count.saturating_inc();
```

Comment:
if a few pools migrate for some reason, what are the side effects of having them? the `StakeStrategyType` is global so that may be an issue, no? Have you considered keeping the `StakeStrategyType` as a per-pool setting?

Reply:
> what are the side effects of having them? 

I guess the `post_upgrade` will fail as the failed pool balances will not be 0. So it should not happen in the runtime.

Reply:
There is a test that checks if a pool fails to migrate, the members still cannot withdraw. (since they don't exist in delegators map in delegation staking pallet).

But this is the refactor we were discussing last week. To recap, I am contemplating whether to make another refactor to:
- Keep a storage map per pool if the migration has happened.
- Make pallet support both `TransferStake` and `DelegateStake` at the same time. We could even separate out `DelegationInterface` and pass it separately providing an impl for `()` if the runtime wants to support only `TransferStake`. Have to try it out (in a separate branch) to see how this ends up.
 - We avoid doing one time migration, but pool can permissionlessly migrate to `DelegateStake` and the appropriate migration flag is updated.
 
## Reasons not to do this refactor?
- We might want to just clean up `TransferStake` completely at some point. If we can ensure that there are no footguns that would corrupt the state, we can argue that the code is fine in its current form and we are probably over engineering.

I will make a note of possible footguns in the main description and if it can lead to issues. The only big footgun I can think off top of my head is if the provided adapter to the pallet is `DelegateStake` but no migration was done to convert pools to `DelegateStake` (or some pools failed to migrate). This would lead to all member calls failing for existing pools since these accounts don't exist in `pallet-delegation-staking`. (scenario covered in test)

@georgepisaltu your opinion here might be valuable too.

Reply:
There is a test that checks if a pool fails to migrate, the members still cannot withdraw. (since they don't exist in delegators map in delegation staking pallet).

I have refactored the storage to be used only to make sure a migration of strategy type has occurred and to avoid redoing the migration.

Reply:
> I have refactored the storage to be used only to make sure a migration of strategy type has occurred and to avoid redoing the migration.

So basically, the pools that failed to migrate will be active but the members cannot migrate to the new strategy right? Will the failed pools works as expected with the transfer stake strat a i.e. can we have different strategies working in at the same time? (I think that would be great!)

Reply:
> Will the failed pools works as expected with the transfer stake strat 

No they will be kinda blocked (won't be able to do anything). I did try to support both adapters at the same time and have it configurable per pool, but honestly I felt it is adding too much complexity for something we don't really want to do (support both adapters at the same time). Ideal path is, we completely remove `TransferStake` at some point since `DelegateStake` is (unarguably) better than the previous strategy.

Also, if we do the try runtime upgrade tests properly, I think we should feel confident nothing would fail. The only thing we should do defensively (imo) here is to make sure nothing harmful happens if there are some failed migrations in the unexpected case.